File: DFS.js
// This is an application of the DFS algorithm in JavaScript
// To run this, you could visit this link: https://www.programiz.com/online-compiler/3jRfcdSdZvHFw
// You could also run this in the console of any browser
// You cold also run this in your Node environment if you have Node installed

let graph = {
    0: [1, 2],
    1: [3, 4],
    2: [5],
    3: [],
    4: [5],
    5: []
};
const graphSize = Object.keys(graph).length;

let state = new Array(graphSize).fill("new");
let dfsCounter = 0, finishCounter = 0;
let dfsNumbers = [], finishNumbers = [];

const dfs = (v) => {
    state[v] = "active";
    dfsNumbers[v] = dfsCounter;
    dfsCounter++;

    graph[v].forEach(edge => {
        if(state[edge] == "new") {
            dfs(edge);
        }
    })

    state[v] = "finished";
    finishNumbers[v] = finishCounter;
    finishCounter++;
}

for (let i = 0; i < graphSize; i++) {
    if (state[i] == "new") {
        dfs(i);
    }
}

console.log("States: ", state);
console.log("DFS Numbers: ", dfsNumbers);
console.log("Finish Numbers: ", finishNumbers);

// Thank you for reading

File: DFS.cpp
// For quick testing, visit this link https://www.programiz.com/online-compiler/6VrRF3dtQGJk9

#include <iostream>
#include <vector>
#include <map>

using namespace std;

map<int, vector<int>> graph = {
    {0, {1, 2}},
    {1, {3, 4}},
    {2, {5}},
    {3, {}},
    {4, {5}},
    {5, {}}
};

const int graphSize = graph.size();

vector<string> state(graphSize, "new");
int dfsCounter = 0, finishCounter = 0;
vector<int> dfsNumbers(graphSize, -1);
vector<int> finishNumbers(graphSize, -1);

void dfs(int v) {
    state[v] = "active";
    dfsNumbers[v] = dfsCounter++;

    for (int edge : graph[v]) {
        if (state[edge] == "new") {
            dfs(edge);
        }
    }

    state[v] = "finished";
    finishNumbers[v] = finishCounter++;
}

int main() {
    for (int i = 0; i < graphSize; ++i) {
        if (state[i] == "new") {
            dfs(i);
        }
    }

    cout << "States: ";
    for (const auto &s : state) {
        cout << s << " ";
    }
    cout << endl;

    cout << "DFS Numbers: ";
    for (const auto &num : dfsNumbers) {
        cout << num << " ";
    }
    cout << endl;

    cout << "Finish Numbers: ";
    for (const auto &num : finishNumbers) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}
